pkg/agent/agent.go

```diff
@@ -146,13 +147,29 @@ func InitializeWithAgentFeatures(centralCfg config.CentralConfig, agentFeaturesC
 
 	if !agent.isInitialized {
 		setupSignalProcessor()
-		// only do the periodic healthcheck stuff if NOT in unit tests and running binary agents
+		// only do the periodic health check stuff if NOT in unit tests and running binary agents
 		if util.IsNotTest() && !isRunningInDockerContainer() {
 			hc.StartPeriodicHealthCheck()
 		}
 
 		if util.IsNotTest() && agent.agentFeaturesCfg.ConnectionToCentralEnabled() {
 			StartAgentStatusUpdate()
+			migration := migrate.NewAttributeMigration(agent.apicClient, agent.cfg)
+			// register the update cache job
+			discoveryCache := newDiscoveryCache(agent.agentResourceManager, false, agent.instanceCacheLock)
+			discoveryCache.Execute()
```

Comment:
@vivekschauhan I moved the call to populate the cache out so that it blocks the other jobs until this is finished. This fixes the race condition. But, I am still unsure if this is what you are looking for because you also mentioned looping through persistent cache in stream mode. This way will populate the cache from central regardless if it is stream or poll mode which makes the persistent cache not that useful.

Reply:
@tsjohns9, for streaming mode what I meant was instead of updating the cache by making API call and executing the migration logic, we should iterate over the cached APIServices and run migration on them.